README.md
# Laravel Auto Head Tags

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

## Description

This package aims to ease the creating of all HTML head tags: meta, links, SEO, Open Graph, Twitter and whatever else you need. 

The package will grab all the info it needs from the data sent to Blade:

```php
return view('welcome', [
    'seo' => [
        'title' => 'Your page title',

        'description' => 'The meta description for the page',

        'urls' => [
            'canonical' => 'https://site.com/the-article-slug'
        ],
    ],

    'twitter' => [
        'handle' => '@opticalcortex'
    ],

    'image' => [
        'url' => 'https://site.com/image.jpg'
    ]
]);
```

To generate, out of the box, this set of tags for you:

```html
<title>Your page title</title>\n
<meta charset="utf-8" />\n
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />\n
<meta http-equiv="X-UA-Compatible" content="IE=Edge" />\n
<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1" />\n
<meta name="mobile-web-app-capable" content="yes" />\n
<meta name="apple-mobile-web-app-capable" content="yes" />\n
<meta name="apple-mobile-web-app-title" content="Your page title" />\n
<meta name="description" content="The meta description for the page" />\n
<meta name="msapplication-TileColor" content="#ffffff" />\n
<meta name="msapplication-config" content="/favicon/browserconfig.xml" />\n
<meta name="theme-color" content="#ffffff" />\n
<meta itemprop="description" content="The meta description for the page" />\n
<meta property="og:type" content="website" />\n
<meta property="og:title" content="Your page title" />\n
<meta property="og:site_name" content="Laravel" />\n
<meta property="og:description" content="The meta description for the page" />\n
<meta property="og:url" content="https://site.com/the-article-slug" />\n
<meta property="og:image:width" content="1200" />\n
<meta property="og:image:height" content="630" />\n
<meta name="twitter:title" content="Your page title" />\n
<meta name="twitter:site" content="@opticalcortex" />\n
<meta name="twitter:description" content="The meta description for the page" />\n
<meta name="twitter:url" content="https://site.com/the-article-slug" />\n
<meta name="twitter:card" content="summary_large_image" />\n
<meta name="twitter:creator" content="@opticalcortex" />\n
<link rel="manifest" href="/favicon/site.webmanifest" />\n
<link rel="icon" href="/favicon/favicon.ico" />\n
<link rel="icon" type="image/png" sizes="32x32" href="/favicon/favicon-32x32.png" />\n
<link rel="icon" type="image/png" sizes="16x16" />\n
<link rel="apple-touch-icon" sizes="180x180" href="/favicon/apple-touch-icon.png" />\n
<link rel="mask-icon" href="/favicon/safari-pinned-tab.svg" color="#1d1d1d" />\n
<link rel="shortcut icon" href="/favicon/favicon.ico" />\n
<link rel="canonical" href="https://site.com/the-article-slug" />
```

[The list of tags is larger](https://github.com/area17/laravel-auto-head-tags/blob/master/config/laravel-auto-head-tags.yaml), but if you don't provide enough information to create those tags, they won't be created.

## Install

### Via Composer

```bash
composer require area17/laravel-auto-head-tags
```

### Publish the config file

```bash
php artisan vendor:publish --provider="A17\LaravelAutoHeadTags\ServiceProvider"
```

# Using

## Add to your blade template

Add the tag `@head` to your main template:

```html
<!DOCTYPE html>
<html lang="{{ locale() }}">
    <head>
        @head

        ...
    </head>

    ...
</html>
```

This directive is also configurable, just publish the configuration and change the `blade.directive`.

## Configuring 

All available tags are on the config file: `config/laravel-auto-head-tags.yaml`, and everything is configurable. This is an extract of the tags section:

```yaml
# TAGS
tags:
    # META

    # HEAD
    meta:
    - charset: "{head.charset}|utf-8"
    - name: viewport
      content: "{head.viewport}|width=device-width, initial-scale=1.0, minimum-scale=1"
    - name: author
      content: "{head.author}"
    - http-equiv: "X-UA-Compatible"
      content: "{head.httpEquiv}|IE=Edge"
    - name: mobile-web-app-capable
      content: "{head.mobile-web-app-capable}|yes"
    - name: apple-mobile-web-app-capable
      content: "{head.apple-mobile-web-app-capable}|yes"
    - name: apple-mobile-web-app-title
      content: "{head.apple-mobile-web-app-title}|{og.title}|{$config.app.name}"
```

You can define macros to access Blade data, using `{}`, and the package can resolve them using the "dot" notation for arrays:

```yaml
content: "{head.author}"
```

You can define as many fallbacks you need for those macros:

```yaml
title: "{title}|{head.title}|{seo.title}|{og.title}|{$config.app.name}"
```

And you can also access data from the Laravel config:

```yaml
content: "{$config.app.name}"
```

If it's required to generate more than one URL for a single tag definition, there's `loop` concept:

```yaml
- rel: canonical
  href: "{seo.urls.canonical}"
- rel: alternate
  href: "{seo.urls.hreflang:value}"
  hreflang: "{seo.urls.hreflang:key}"
```

If `seo.urls.hreflang` is an array made of locales (`key`) and URLs (`value`), this configuration will generate these tags:

```html
<link rel="canonical" href="http://site.com/fr/events/event-slug" />
<link rel="alternate" href="http://site.com/fr/evenements/event-slug" hreflang="fr" />
<link rel="alternate" href="http://site.com/en/events/event-slug" hreflang="en" />
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

```bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email antonio@area17.com instead of using the issue tracker.

## Credits

- [AREA 17](https://github.com/area17)
- [Antonio Ribeiro][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/area17/laravel-auto-head-tags.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://scrutinizer-ci.com/g/area17/laravel-auto-head-tags/badges/build.png?b=master&style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/area17/laravel-auto-head-tags.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/area17/laravel-auto-head-tags.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/area17/laravel-auto-head-tags.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/area17/laravel-auto-head-tags
[link-travis]: https://travis-ci.com/github/area17/laravel-auto-head-tags
[link-scrutinizer]: https://scrutinizer-ci.com/g/area17/laravel-auto-head-tags/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/area17/laravel-auto-head-tags
[link-downloads]: https://packagist.org/packages/area17/laravel-auto-head-tags
[link-author]: https://github.com/antonioribeiro
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `laravel-auto-head-tags` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/HeadTest.php
<?php

namespace A17\LaravelAutoHeadTags\Tests;

use A17\LaravelAutoHeadTags\Head;
use Illuminate\Support\Facades\Blade;

class HeadTest extends TestCase
{
    public function testCanLoadConfig()
    {
        $this->assertEquals(
            '$config',
            config('laravel-auto-head-tags.config.key')
        );
    }

    public function testCanCompileBladeExtension()
    {
        $bladeSnippet = '@head';

        $expectedCode =
            '<?php echo (new A17\LaravelAutoHeadTags\Head($__data))->render(); ?>';

        $this->assertEquals($expectedCode, Blade::compileString($bladeSnippet));
    }

    public function testCanRenderHead()
    {
        $rendered = $this->head->render();

        $this->assertStringContainsString(
            '<meta name="twitter:title" content="OG - Your page title" />',
            $rendered
        );
    }
}


tests/TestCase.php
<?php

namespace A17\LaravelAutoHeadTags\Tests;

use A17\LaravelAutoHeadTags\Head;
use A17\LaravelAutoHeadTags\ServiceProvider;
use Orchestra\Testbench\TestCase as OrchestraTestCase;

class TestCase extends OrchestraTestCase
{
    /**
     * @var \A17\LaravelAutoHeadTags\Head
     */
    protected $head;

    protected $data = [
        'seo' => [
            'title' => 'SEO - Your page title',

            'description' => 'The meta description for the page',

            'urls' => [
                'canonical' => 'https://site.com/the-article-slug',

                'hreflang' => [
                    'fr' => 'https://site.com/fr/the-article-slug',
                    'en' => 'https://site.com/en/the-article-slug'
                ]
            ]
        ],

        'twitter' => [
            'handle' => '@opticalcortex'
        ],

        'image' => [
            'url' => 'https://site.com/image.jpg'
        ],

        'og' => [
            'title' => 'OG - Your page title',

            'site-name' => 'App name',

            'locale' => ['current' => 'fr', 'alternate' => ['fr', 'en']],

            'type' => null,

            'description' => 'Description',

            'url' => 'https://site.com/article-slug',

            'image' => [
                'url' => 'https://site.com/article-slug/image.jpg',

                'secure-url' => 'https://site.com/article-slug/image.jpg',

                'alt' => 'Image alt',

                'type' => 'image/jpeg',

                'width' => '800',

                'height' => '600'
            ],

            'video' => [
                'url' => 'http://site.com/article-slug/video.mp4',

                'secure-url' => 'https://site.com/article-slug/video.mp4',

                'alt' => 'Video alt',

                'type' => 'text/html'
            ]
        ]
    ];

    public function setUp(): void
    {
        parent::setUp();

        $this->head = new Head($this->data);
    }

    protected function getPackageProviders($app)
    {
        return [ServiceProvider::class];
    }
}


tests/ExampleTest.php
<?php

declare(strict_types=1);

namespace A17\TwillHead;

class ExampleTest extends \PHPUnit\Framework\TestCase
{
    /**
     * Test that true does in fact equal true
     */
    public function testTrueIsTrue()
    {
        $this->assertTrue(true);
    }
}


src/Head.php
<?php

namespace A17\LaravelAutoHeadTags;

use Illuminate\Support\Str;
use Illuminate\Support\Arr;
use Illuminate\Support\Collection;
use A17\LaravelAutoHeadTags\Behaviors\Macro;
use A17\LaravelAutoHeadTags\Behaviors\Config;

class Head
{
    use Macro, Config;

    /**
     * @var \Illuminate\Support\Collection
     */
    protected $data;

    /**
     * @var \Illuminate\Support\Collection
     */
    protected $config;

    /**
     * @var boolean
     */
    protected $firstLineRendered = false;

    /**
     * Head constructor.
     *
     * @param array|\Illuminate\Support\Collection $data
     */
    public function __construct($data)
    {
        $this->setData($data);

        $this->setConfig(config()->all());
    }

    /**
     * @param \Illuminate\Support\Collection $properties
     * @return \Illuminate\Support\Collection
     */
    protected function fillUpNonArrayProperties(Collection $properties)
    {
        $traversable = $properties->first(function ($property) {
            return is_traversable($property);
        });

        return $properties->map(function ($value) use ($traversable) {
            if (is_traversable($value)) {
                return $value;
            }

            return $traversable->map(function () use ($value) {
                return $value;
            });
        });
    }

    /**
     * @param \Illuminate\Support\Collection $properties
     * @return \Illuminate\Support\Collection
     */
    protected function generatePropertiesArray(Collection $properties)
    {
        if (!$this->propertiesContainsLoops($properties)) {
            return collect([$properties]);
        }

        $properties = $this->fillUpNonArrayProperties(
            $properties->map(function ($property) {
                return $this->explodePropertyLoop($property);
            })
        );

        $result = [];

        $properties->each(function (Collection $values, $propertyName) use (
            &$result
        ) {
            $values->each(function ($value, $key) use (
                $propertyName,
                &$result
            ) {
                Arr::set($result, "{$key}.{$propertyName}", $value);
            });
        });

        return collect($result);
    }

    /**
     * @param string $tagName
     * @param array|\Illuminate\Support\Collection $tags
     * @return \Illuminate\Support\Collection
     */
    protected function generateTagsWithProperties($tagName, $tags): Collection
    {
        return collect($tags)
            ->map(function ($properties) use ($tagName) {
                return $this->generatePropertiesArray($properties)->map(
                    function ($properties) use ($tagName) {
                        return $this->generateTag($tagName, $properties);
                    }
                );
            })
            ->filter();
    }

    /**
     * @param string $tagName
     * @param array|string|\Illuminate\Support\Collection $tags
     * @return \Illuminate\Support\Collection
     */
    protected function generateTagsWithoutProperties($tagName, $tags): Collection
    {
        return collect(
            $this->generateTag(
                $tagName,
                $tags,
                'tag.without_properties_format',
                'tag.value_only_format'
            )
        )->filter();
    }

    /**
     * @return string
     */
    protected function getFallbackDelimiter()
    {
        return $this->config('delimiters.fallback');
    }

    /**
     * @return string
     */
    protected function getIndent()
    {
        if (!$this->firstLineRendered) {
            $this->firstLineRendered = true;

            return '';
        }

        return str_repeat(' ', $this->config('tag.indent.spaces')) .
            str_repeat("\t", $this->config('tag.indent.tabs'));
    }

    /**
     * @return string
     */
    public function render()
    {
        return $this->config('tags')
            ->map(function ($properties, $tag) {
                return $this->generateTags($tag, $properties);
            })
            ->flatten()
            ->filter()
            ->implode($this->config('tag.line_break'));
    }

    /**
     * @param string $tagName
     * @param string|array|\Illuminate\Support\Collection $tags
     * @return \Illuminate\Support\Collection
     */
    public function generateTags($tagName, $tags)
    {
        return is_string($tags)
            ? $this->generateTagsWithoutProperties($tagName, $tags)
            : $this->generateTagsWithProperties($tagName, $tags);
    }

    /**
     * @param string $tagName
     * @param string|array|\Illuminate\Support\Collection $value
     * @param string $tagFormat
     * @param string $propertiesFormat
     * @return string|null
     */
    public function generateTag(
        $tagName,
        $value,
        $tagFormat = 'tag.with_properties_format',
        $propertiesFormat = 'tag.property_format'
    ) {
        $value = $this->generateProperties(
            $tagName,
            $value,
            $propertiesFormat
        )->implode(' ');

        if (filled($value)) {
            return $this->getIndent() .
                str_replace(
                    ['{tagName}', '{value}'],
                    [$tagName, $value],

                    $this->config($tagFormat)
                );
        }

        return null;
    }

    /**
     * @param string $tagName
     * @param string|array|\Illuminate\Support\Collection $properties
     * @param string $format
     * @return \Illuminate\Support\Collection
     */
    public function generateProperties(
        $tagName,
        $properties,
        $format = 'tag.property_format'
    ) {
        $properties = collect($properties)->map(function ($value, $key) use (
            $format
        ) {
            $value = $this->generateValue($value);

            return [
                'key' => $key,
                'value' => $value,
                'rendered' => filled($value)
                    ? str_replace(
                        ['{propertyName}', '{value}'],
                        [$key, $value],
                        $this->config($format)
                    )
                    : null
            ];
        });

        if (
            $this->metaIsBlank($properties, $tagName) ||
            $this->linkIsBlank($properties, $tagName)
        ) {
            return collect();
        }

        return $properties->map(function ($property) {
            return $property['rendered'];
        });
    }

    /**
     * @param \Illuminate\Support\Collection $properties
     * @param string $tagName
     * @return bool
     */
    protected function linkIsBlank($properties, $tagName): bool
    {
        return $tagName === 'link' &&
            isset($properties['href']) &&
            blank($properties['href']['value']);
    }

    /**
     * @param \Illuminate\Support\Collection $properties
     * @param string $tagName
     * @return bool
     */
    protected function metaIsBlank($properties, $tagName): bool
    {
        return $tagName === 'meta' &&
            isset($properties['content']) &&
            blank($properties['content']['value']);
    }

    /**
     * @param string $value
     * @return array|mixed|null
     */
    public function generateValue($value)
    {
        return collect(explode($this->getFallbackDelimiter(), $value))
            ->map(function ($value) {
                return $this->makeValue($value);
            })
            ->first(function ($value) {
                return filled($value);
            });
    }

    /**
     * @param array $config
     * @return Head
     */
    public function setConfig(array $config): Head
    {
        $this->config = to_collection_recursive($config);

        return $this;
    }

    /**
     * @param array|\Illuminate\Support\Collection $data
     * @return Head
     */
    public function setData($data): Head
    {
        $this->data = collect($data);

        return $this;
    }

    /**
     * @param string $value
     * @return array
     */
    public function splitDefault($value)
    {
        $delimiter = $this->getFallbackDelimiter();

        if (!Str::contains($value, $delimiter)) {
            return [$value, null];
        }

        return [
            Str::before($value, $delimiter),
            Str::after($value, $delimiter)
        ];
    }

    /**
     * @param string $value
     * @return array|mixed|null
     */
    public function makeValue($value)
    {
        if (blank($macro = $this->extractMacro($value))) {
            return $value;
        }

        return $this->makeValueFromMacro($macro);
    }
}


src/Behaviors/Macro.php
<?php

namespace A17\LaravelAutoHeadTags\Behaviors;

use Illuminate\Support\Str;
use Illuminate\Support\Arr;
use Illuminate\Support\Collection;

trait Macro
{
    abstract public function config($key = null);

    abstract protected function getConfigKey();

    /**
     * @param string $property
     * @return bool
     */
    protected function containsMacro($property)
    {
        return filled($this->extractMacro($property));
    }

    /**
     * @param string $property
     * @return string|null
     */
    protected function extractAndCleanMacro($property)
    {
        if (blank($macro = $this->extractMacro($property))) {
            return null;
        }

        return $this->removeConfigKeyFromMacro($macro);
    }

    /**
     * @param string $property
     * @return string|null
     */
    protected function extractMacroFromLoopMacro($property)
    {
        $macro = $this->extractAndCleanMacro($property);

        if (!Str::contains($macro, $this->getLoopDelimiter())) {
            return null;
        }

        return Str::beforeLast($macro, $this->getLoopDelimiter());
    }

    /**
     * @param string $property
     * @return string|null
     */
    protected function extractVarFromLoopMacro($property)
    {
        $macro = $this->extractAndCleanMacro($property);

        if (!Str::contains($macro, $this->getLoopDelimiter())) {
            return null;
        }

        return Str::afterLast($macro, $this->getLoopDelimiter());
    }

    /**
     * @param string $value
     * @return string|null
     */
    protected function extractMacro($value)
    {
        preg_match_all($this->getMacroRegexParser(), $value, $matches);

        if (blank($matches[1][0] ?? null)) {
            return null;
        }

        return $matches[1][0];
    }

    /**
     * @return string
     */
    protected function getMacroRegexParser()
    {
        return $this->config('macro.regex_parser');
    }

    /**
     * @param string $macro
     * @return array|mixed|null
     */
    protected function makeValueFromMacro($macro)
    {
        return Str::contains($macro, '.')
            ? $this->makeValueFromMacroArray($macro)
            : $this->makeValueFromMacroVariable($macro);
    }

    /**
     * @param string $macro
     * @return mixed|null
     */
    protected function makeValueFromMacroVariable($macro)
    {
        return $this->data[$macro] ?? null;
    }

    /**
     * @param string $macro
     * @return string
     */
    protected function removeConfigKeyFromMacro($macro): string
    {
        $macro = Str::startsWith($macro, $this->getConfigKey())
            ? Str::after($macro, $this->getConfigKey())
            : $macro;

        return $macro;
    }

    /**
     * @param string $variable
     * @return array|mixed
     */
    public function makeValueFromMacroArray($variable)
    {
        $keys = Str::after($variable, '.');

        $variable = Str::before($variable, '.');

        $value =
            $variable === $this->getConfigKey()
                ? $this->config
                : $this->data[$variable] ?? collect();

        if (is_traversable($value)) {
            return Arr::get(to_array($value), $keys);
        }

        return $value;
    }

    /**
     * @param string $property
     * @return bool
     */
    protected function containsLoop($property)
    {
        return filled($this->extractVarFromLoopMacro($property));
    }

    /**
     * @param string $property
     * @return string|\Illuminate\Support\Collection
     */
    protected function explodePropertyLoop($property)
    {
        if (!$this->containsLoop($property)) {
            return $property;
        }

        $key = $this->extractVarFromLoopMacro($property);

        $macro = $this->extractMacroFromLoopMacro($property);

        if (blank($macro)) {
            return $property;
        }

        return collect($this->makeValueFromMacro($macro))
            ->map(function ($value, $key) {
                return compact('value', 'key');
            })
            ->map(function ($values) use ($key) {
                return $values[$key];
            });
    }

    /**
     * @return string
     */
    protected function getLoopDelimiter()
    {
        return $this->config('delimiters.loop');
    }

    /**
     * @param \Illuminate\Support\Collection $properties
     * @return mixed
     */
    protected function propertiesContainsLoops(Collection $properties)
    {
        return $properties->reduce(function ($keep, $property) {
            return $keep ||
                ($this->containsMacro($property) &&
                    $this->containsLoop($property));
        }, false);
    }
}


src/ServiceProvider.php
<?php

namespace A17\LaravelAutoHeadTags;

use PragmaRX\Yaml\Package\Yaml;
use Illuminate\Support\Facades\Blade;
use Illuminate\Support\ServiceProvider as IlluminateServiceProvider;

class ServiceProvider extends IlluminateServiceProvider
{
    /**
     * Boot
     */
    public function boot()
    {
        $this->publishConfig();

        $this->configureBlade();
    }

    /**
     * Register
     */
    public function register()
    {
        $this->mergeConfig();
    }

    /**
     * Publish config
     */
    public function publishConfig()
    {
        $this->publishes(
            [
                __DIR__ .
                '/../config/laravel-auto-head-tags.yaml' => config_path(
                    'laravel-auto-head-tags.yaml'
                )
            ],
            'config'
        );
    }

    /**
     * Configure blade
     */
    public function configureBlade()
    {
        Blade::directive(
            config('laravel-auto-head-tags.blade.directive'),
            function () {
                return '<?php echo (new A17\LaravelAutoHeadTags\Head($__data))->render(); ?>';
            }
        );
    }

    /**
     * Merge config
     */
    public function mergeConfig()
    {
        $app = config_path('laravel-auto-head-tags.yaml');

        $package = __DIR__ . '/../config/laravel-auto-head-tags.yaml';

        (new Yaml())->loadToConfig(
            file_exists($app) ? $app : $package,
            'laravel-auto-head-tags'
        );
    }
}


src/Support/helpers.php
<?php

use Illuminate\Support\Str;

if (!function_exists('to_collection_recursive')) {
    /**
     * @param array $array
     * @return \Illuminate\Support\Collection
     */
    function to_collection_recursive($array)
    {
        return collect($array)->map(function ($item) {
            if (is_array($item)) {
                return collect(to_collection_recursive($item));
            }

            return $item;
        });
    }
}

if (!function_exists('is_traversable')) {
    /**
     * @param mixed $subject
     * @return bool
     */
    function is_traversable($subject)
    {
        return is_array($subject) || $subject instanceof ArrayAccess;
    }
}

if (!function_exists('to_array')) {
    function to_array($collection)
    {
        if (!is_traversable($collection)) {
            return $collection;
        }

        if (is_array($collection)) {
            $collection = collect($collection);
        } else {
            $collection = collect($collection->toArray());
        }

        $collection = $collection->map(function ($item) {
            if (is_traversable($item)) {
                return to_array($item);
            }

            return $item;
        });

        if (keys_are_all_numeric($collection)) {
            $collection = $collection->values();
        }

        return $collection->toArray();
    }
}

if (!function_exists('keys_are_all_numeric')) {
    function keys_are_all_numeric($array)
    {
        return collect($array)
            ->keys()
            ->reduce(function ($keep, $key) {
                return $keep && is_integer($key);
            }, true);
    }
}


config/twill-head.php
<?php

return [
    'meta' => [
        ['charset' => 'utf-8'],
        ['http-equiv' => 'X-UA-Compatible', 'content' => 'IE=Edge'],

        ['property' => 'og:title', 'content' => '{ogTitle}'],
        ['property' => 'og:title', 'content' => '{ogTitle}'],
        ['property' => 'og:title', 'content' => '{ogTitle}'],
        ['property' => 'og:site_name', 'content' => '{ogSiteName}'],
        ['property' => 'og:locale', 'content' => '{ogLocale}'],
        ['property' => 'og:type', 'content' => 'website'],
        ['property' => 'og:description', 'content' => '{ogDescription}'],
        ['property' => 'og:url', 'content' => '{canonicalUrl}'],
        ['property' => 'og:image', 'content' => '{ogImage}'],
        ['property' => 'og:image:width', 'content' => '1200'],
        ['property' => 'og:image:height', 'content' => '630'],

        ['name' => 'viewport', 'content' => 'width=device-width, initial-scale=1.0, minimum-scale=1'],
        ['name' => 'twitter:title', 'content' => '{twitterTitle}'],
        ['name' => 'twitter:site', 'content' => '{twitterHandle}'],
        ['name' => 'description', 'content' => '{description}'],
        ['name' => 'twitter:description', 'content' => '{twitterDescription}'],
        ['name' => 'twitter:card', 'content' => 'summary_large_image'],
        ['name' => 'twitter:image', 'content' => '{twitterImage}'],
        ['name' => 'twitter:url', 'content' => '{canonicalUrl}'],
        ['name' => 'msapplication-TileColor', 'content' => '#ffffff'],
        ['name' => 'msapplication-config', 'content' => '/dist/images/favicons/browserconfig.xml'],
        ['name' => 'theme-color', 'content' => '#ffffff'],

        ['itemprop' => 'name', 'content' => '{title}'],
        ['itemprop' => 'description', 'content' => '{description}'],
        ['itemprop' => 'image', 'content' => '{seoImage }'],
    ],

    'link' => [
        ['rel' => 'canonical', 'href' => '{canonicalUrl}'],
        ['rel' => 'apple-touch-icon', 'sizes' => '180x180', 'href' => '/dist/images/favicons/apple-touch-icon.png'],
        ['rel' => 'icon" type="image/png', 'sizes' => '32x32', 'href' => '/dist/images/favicons/favicon-32x32.png'],
        ['rel' => 'icon" type="image/png', 'sizes' => '16x16', 'href' => '/dist/images/favicons/favicon-16x16.png'],
        ['rel' => 'manifest', 'href' => '/dist/images/favicons/site.webmanifest'],
        ['rel' => 'mask-icon', 'href' => '/dist/images/favicons/safari-pinned-tab.svg" color="#000000'],
        ['rel' => 'shortcut icon', 'href' => '/dist/images/favicons/favicon.ico'],
        ['rel' => 'icon', 'href' => '/dist/images/favicons/favicon.svg'],
    ]
];
